### PR TITLE
Mitigate tob-5 by making a v2 of client_propose_tx endpoint

### PR DESCRIPTION
**Pull request by garbageslam**
TOB-MCCT-5 is about a "token id oracle" described by TOB during review of confidential tokens design. The idea is that the consensus enclave fee map can be changed by the untrusted side, by restarting the node. If the attacker controls a node and is able to get the user to submit their Tx once with one fee map, and once with another, then the priority value leaks information about what the token id of the Tx is.

In principle they could do this without the user knowing -- they would restart the node and it would not be able to attest to the peers after changing the fee map, but it could still accept Tx proposals and see the WellFormedTxContext which conatins the priority number. Then they turn the node back to normal. The user's app retries when their transaction doesn't land, and the same Tx comes back to their node, now with the regular fee map.

This requires having root on a consensus node, but the threat model is that attackers with root on the consensus node should not be able to see the user's token id.

We view this as a time-of-check-vs-time-of-use ([TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use)) bug, and we try to fix it a way that those bugs are sometimes fixed. The user, who checked the fee map, also makes an assertion about what the fee-map is when they try to use the system (by submitting a Tx). Before acting on the data (which the attacker has potentially tampered with), the enclave checks if the user's assertion is correct, and rejects the Tx without creating a WellFormedTxContext if the assertion is not correct. (This is sort of like dealing with races when manipulating atomic variables by using compare-and-swap.)

It should be straightforward to add this check, but unfortunately, the status quo `client_propose_tx` has a design flaw. It does not have a `Request` protobuf as the RPC argument, the argument it takes is directly an (encrypted) Tx. So, we cannot add any additional data to that protobuf in a non-breaking way without adding things to `Tx`. And that is untenable in this case because it adds complexity in sensitive areas of the code. We really just want to add this check next to the `Tx`.

To fix this, we introduce a `ClientProposeTxRequest` object, which contains a `Tx`, and the client's minimum fee map digest. We introduce `client_propose_tx_v2` which encrypts this `Request` object instead of a naked `Tx`. This will come in handy if we want to add anything else like this in the future.

Clients will not be able to use `client_propose_tx_v2` until after the V3 service is live, because existing consensus nodes don't have this endpoint. However, in the next release of clients after the v3 release is live, they could all start using `client_propose_tx_v2` and we could deprecate / remove `client_propose_tx`.

I made the fog-sample-paykit exercise the `client_propose_tx_v2` route, since the test client uses that and so we can see that it is working in CD without making `mobilecoind` or anything else use it yet.

I am not sure if this change requires an MCIP. It is a protocol / api change, but not a big one, and users are unlikely to notice. If it does require an MCIP we can write one pretty easily. Let me know what you think.

Also, lmk if you think there's a better approach than the route I took here.

### CODE REVIEW

File: consensus/enclave/api/src/lib.rs [hunk 1]
```diff
@@ -246,6 +247,22 @@ pub struct FormBlockInputs {
     pub mint_txs_with_config: Vec<(MintTx, MintConfigTx, MintConfig)>,
 }
 
+/// The schema expected to be encrypted for client_tx_propose_v2
+#[derive(Clone, Message)]
+pub struct ClientProposeTxRequest {
```
Comment (garbageslam) [inline consensus/enclave/api/src/lib.rs:252]: This is the request object that clients use to propose Tx's in the new `client_propose_tx_v2`.

File: consensus/enclave/api/src/lib.rs [hunk 2]
```diff
@@ -311,14 +328,41 @@ pub trait ConsensusEnclave: ReportableEnclave {
...
+    ///
+    /// Arguments:
+    /// * An EnclaveMessage which encrypts a `ClientProposeTxRequest` protobuf
+    ///
+    /// Returns:
+    /// * A Tx context including "locally encrypted" version of Tx and some extracted
+    ///   metadata, or
+    /// * A serialization / cryptography error, or another mismatch error
+    ///   e.g. minimum_fee_map_hash or (some-day) chain-id error
+    fn client_tx_propose_v2(&self, msg: EnclaveMessage<ClientSession>) -> Result<TxContext>;
```
Comment (garbageslam) [inline consensus/enclave/api/src/lib.rs:352]: we can't just fold these changes into `client_tx_propose` because we need to make a breaking change to the schema in order to insert another layer where we can add things. Making a new v2 version means that the old API continues to work and support clients when we drop in the new server.

File: consensus/enclave/impl/src/lib.rs [hunk 3]
```diff
@@ -610,27 +638,23 @@ impl ConsensusEnclave for SgxConsensusEnclave {
...
-            tx_hash,
-            highest_indices,
-            key_images,
-            output_public_keys,
-        })
+        if &req.fee_map_digest[..] != &self
+            .blockchain_config
+            .get()
+            .ok_or(Error::NotInitialized)?.canonical_fee_map_digest()[..] {
+            return Err(Error::FeeMapDigestMismatch);
```
Comment (garbageslam) [inline consensus/enclave/impl/src/lib.rs:654]: this is where we actually check if the hash matches

File: fog/sample-paykit/src/client.rs [hunk 4]
```diff
@@ -183,17 +198,22 @@ impl Client {
     /// transaction.
     pub fn send_transaction(&mut self, transaction: &Tx) -> Result<u64> {
         let start_time = std::time::SystemTime::now();
-        let block_count = self.consensus_service_conn.propose_tx(transaction)?;
+        let fee_map = self.get_latest_fee_map()?;
+        let tx_ok_data = self.consensus_service_conn.propose_tx_v2(transaction, &fee_map).map_err(|err| {
+            // Clear fee map in case this was the problem
+            self.fee_map = None;
```
Comment (garbageslam) [inline fog/sample-paykit/src/client.rs:204]: Note: in a real client, to defend against the attack, what you should actually do is check the fee map with multiple consensus nodes. if they don't agree then something bad is happening to the network, either maliciously or not. nodes that dont agree about the fee map necessarily fail to consense to one another.

Comment (garbageslam) [reply fog/sample-paykit/src/client.rs:204]: I guess I should add a code comment like this

File: connection/test-utils/src/blockchain.rs [hunk 5]
```diff
@@ -133,6 +133,14 @@ impl<L: Ledger + Sync> UserTxConnection for MockBlockchainConnection<L> {
         self.proposed_txs.push(tx.clone());
         Ok(self.ledger.num_blocks().unwrap())
     }
+
+    fn propose_tx_v2(&mut self, tx: &Tx, _fee_map: &FeeMap) -> ConnectionResult<TxOkData> {
+        self.proposed_txs.push(tx.clone());
+        Ok(TxOkData {
+            block_count: self.ledger.num_blocks().unwrap(),
+            block_version: 0,
```
Comment (eranrund) [inline connection/test-utils/src/blockchain.rs:141]: Nit: `BlockVersion::MAX` might be more correct here

File: connection/src/thick.rs [hunk 6]
```diff
@@ -435,6 +436,51 @@ impl<CP: CredentialsProvider> UserTxConnection for ThickClient<CP> {
...
+
+        // Don't leave the plaintext serialization floating around
+        let tx_plaintext = SecretVec::new(encode(&req));
+        let tx_ciphertext =
+            enclave_connection.encrypt(&[], tx_plaintext.expose_secret().as_ref())?;
+        msg.set_data(tx_ciphertext);
+
+        let resp = self.authenticated_attested_call(|this, call_option| {
+            this.consensus_client_api_client
+                .client_tx_propose_async_opt(&msg, call_option)
```
Comment (eranrund) [inline connection/src/thick.rs:468]: Shouldn't this be the v2 api call?

File: consensus/enclave/api/src/lib.rs (hunk 1 above)
Comment (eranrund) [inline consensus/enclave/api/src/lib.rs:252]: Might be worth renaming this to `ClientProposeTxV2Request` to match the service function call name

File: fog/sample-paykit/src/client.rs [hunk 7]
```diff
@@ -173,6 +176,18 @@ impl Client {
         self.tx_data.get_latest_block_version()
     }
 
+    /// Get the latest fee map
+    pub fn get_latest_fee_map(&mut self) -> Result<FeeMap> {
```
Comment (eranrund) [inline fog/sample-paykit/src/client.rs:180]: Looks like this won't necessarily get the latest fee map since once it caches it, it no longer tries again. So that name is misleading.

**Review by eranrund — COMMENTED** [review]
Nice, that looks like it's going in the right direction.
However, it appears there are zero tests for this. Would it be possible to test at least the enclave part?
We likely need to incorporate this into mobilecoind, integration tests, etc as well...

**Comment by garbageslam** [comment]
The other thing I'm not sure about is, maybe I should try to categorize this as a transaction validation error rather than an enclave error. It may be easier for the client to properly handle it if that is done.

**Comment by garbageslam** [comment]
> We likely need to incorporate this into mobilecoind, integration tests, etc as well...

I think we can do this in a separate PR, that can land after branching for v3 perhaps?

If we try to make mobilecoind use this in the v3 release then the mobilecoind for that release can't be in customer hands before the v3 servers are live, which is not usually how we do releases. I guess maybe mobilecoind could switch between calling `client_tx_propose` and `client_tx_propose` based on block version? I didn't think about this before now.

**Comment by garbageslam** [comment]
Here's what I suggest -- we should do a follow-on PR for mobilecoind that uses this, and see that it passes tests, to build confidence in this change.

But we should only merge that for `release/v3.1` and not to `release/v3`

**Comment by jcape** [comment]
It's an eventually breaking API change for users, this will require an MCIP.

**Comment by jcape** [comment]
More to the point, what's wrong with just renaming `Tx` to `TxRequest` and adding the TOB-5 mitigations required?

**Comment by garbageslam** [comment]
> More to the point, what's wrong with just renaming Tx to TxRequest and adding the TOB-5 mitigations required?

This would be the same as adding the fee map hash to the `Tx`.

This would put additional complexity into the Tx, make the Tx hash dependent on this data, and increase the complexity of transaction validation. That's a much larger change that requires touching more sensitive code.

File: consensus/service/src/api/client_api_service.rs [hunk 8]
```diff
@@ -83,15 +84,11 @@ impl ClientApiService {
...
         let tx_context = self.enclave.client_tx_propose(msg.into())?;
-        let mut response = ProposeTxResponse::new();
 
         // Cache the transaction. This performs the well-formedness checks.
         let tx_hash = self.tx_manager.insert(tx_context).map_err(|err| {
             if let TxManagerError::TransactionValidation(cause) = &err {
                 counters::TX_VALIDATION_ERROR_COUNTER.inc(&format!("{:?}", cause));
-                let result = ProposeTxResult::from(cause.clone());
-                response.set_result(result);
-                response.set_err_msg(cause.to_string());
```
Comment (garbageslam) [inline consensus/service/src/api/client_api_service.rs:94]: these lines are removed because this is actually dead code.

this was kind of a head scratcher, but I think this is dead code introduced 2 years ago here: https://github.com/mobilecoinfoundation/mobilecoin/commit/894dd54056b7b9081c306b2fdc9da23f8c18992a

File: consensus/service/src/api/grpc_error.rs [hunk 9]
```diff
@@ -139,6 +139,7 @@ impl From<ConsensusGrpcError> for Result<ProposeTxResponse, RpcStatus> {
         match src {
             ConsensusGrpcError::TransactionValidation(err) => {
                 let mut resp = ProposeTxResponse::new();
+                resp.set_err_msg(err.to_string());
```
Comment (garbageslam) [inline consensus/service/src/api/grpc_error.rs:142]: this is a line that I failed to add to an earlier PR: https://github.com/mobilecoinfoundation/mobilecoin/pull/2501

**Comment by eranrund** [comment]
> > We likely need to incorporate this into mobilecoind, integration tests, etc as well...
> 
> I think we can do this in a separate PR, that can land after branching for v3 perhaps?
> 
> If we try to make mobilecoind use this in the v3 release then the mobilecoind for that release can't be in customer hands before the v3 servers are live, which is not usually how we do releases. I guess maybe mobilecoind could switch between calling `client_tx_propose` and `client_tx_propose_v2` based on block version? I didn't think about this before now.
> 
> Depending on how much other stuff comes up before we are scheduled to branch for v3 I am willing to try to do that.

Good point about the sequencing. But we do need to make sure this API works and for that we need clients that support it, so the block version switching logic might be required. We just need to make sure it really works, so maybe it should assert if block version is 0 since that should no longer happen (my concern is places where we might not be properly including the block version - like the issue we found in the full service  validator earlier this week).

**Comment by garbageslam** [comment]
> > > We likely need to incorporate this into mobilecoind, integration tests, etc as well...
> > 
> > 
> > I think we can do this in a separate PR, that can land after branching for v3 perhaps?
> > If we try to make mobilecoind use this in the v3 release then the mobilecoind for that release can't be in customer hands before the v3 servers are live, which is not usually how we do releases. I guess maybe mobilecoind could switch between calling `client_tx_propose` and `client_tx_propose_v2` based on block version? I didn't think about this before now.
> > Depending on how much other stuff comes up before we are scheduled to branch for v3 I am willing to try to do that.
> 
> Good point about the sequencing. But we do need to make sure this API works and for that we need clients that support it, so the block version switching logic might be required. We just need to make sure it really works, so maybe it should assert if block version is 0 since that should no longer happen (my concern is places where we might not be properly including the block version - like the issue we found in the full service validator earlier this week).

The thing is that we cannot do this block version switching in thick client because thick client has no idea what the block version is. Tx's are structures that may conform to one or more sets of block version rules, and the block version they target is not recorded.

What do you think about this approach: https://github.com/mobilecoinfoundation/mobilecoin/pull/2670#issuecomment-1271096819

So, no clients would be using this new pathway in the wild until 3.1 release, and no block version switching is required, the 3.1 client release would assume this endpoint exists.

I think that's less work overall. We can develop the PR to make those clients do that in a separate PR and when that PR is passing CD we can be sure this PR is working and merge it confidently for v3. But that follow-on PR would not actually merge until v3.1

**Comment by garbageslam** [comment]
My plan right now is to add a unit test that exercises this functionality in `consensus-enclave-impl`, and then do any subsequent work in a follow on PR for mobilecoind.

**Comment by eranrund** [comment]
> What do you think about this approach: [#2670 (comment)](https://github.com/mobilecoinfoundation/mobilecoin/pull/2670#issuecomment-1271096819)

I think that's an acceptable compromise but maybe would be better if we merge that to master and revert it on release/v3?

**Review by nick-mobilecoin — APPROVED** [review]

File: connection/src/thick.rs [hunk 10]
```diff
@@ -435,6 +436,50 @@ impl<CP: CredentialsProvider> UserTxConnection for ThickClient<CP> {
             ))
         }
     }
+
+    fn propose_tx_v2(&mut self, tx: &Tx, fee_map: &FeeMap) -> Result<TxOkData> {
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply connection/src/thick.rs:440]: My hope is that we will delete `propose_tx` in a subsequent client release, and then it will be as DRY as before

File: fog/sample-paykit/src/client.rs [hunk 11]
```diff
@@ -173,6 +176,20 @@ impl Client {
         self.tx_data.get_latest_block_version()
     }
 
+    /// Get the currently cached fee-map
+    ///
+    /// If the cache is empty, then we get fresh data from consensus.
+    pub fn get_fee_map(&mut self) -> Result<FeeMap> {
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply fog/sample-paykit/src/client.rs:182]: I think it's useful in the test-client, I want to use it in the test client to avoid making the operator pass the fee as a parameter. We can make a ticket for that if you like

File: connection/src/thick.rs [hunk 12]
```diff
@@ -435,6 +436,50 @@ impl<CP: CredentialsProvider> UserTxConnection for ThickClient<CP> {
...
+        let req = ClientProposeTxRequestV2 {
+            tx: tx.clone(),
+            fee_map_digest: fee_map.canonical_digest().to_vec(),
+        };
+
+        let mut msg = Message::new();
+        msg.set_channel_id(Vec::from(enclave_connection.binding()));
+
+        // Don't leave the plaintext serialization floating around
+        let tx_plaintext = SecretVec::new(encode(&req));
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply connection/src/thick.rs:461]: Yes, we are not zeroizing everything we should be.
Historically there were problems because Isis' upstream crate did not derive Zeroize on `RistrettoPoint`.

We should make a ticket to fix all the zeroize issues, but this is orthogonal to this PR, and we should not let perfect be the enemy of good.

File: consensus/service/src/api/client_api_service.rs [hunk 13]
```diff
@@ -250,6 +293,54 @@ impl ConsensusClientApi for ClientApiService {
...
+            } else if !(self.is_serving_fn)() {
+                // This node is unable to process transactions (e.g. is syncing its ledger).
+                if let Err(e) = self.enclave.client_discard_message(msg.into()) {
+                    ConsensusGrpcError::Enclave(e).into()
+                } else {
+                    ConsensusGrpcError::NotServing.into()
+                }
+            } else {
+                self.handle_proposed_tx_v2(msg)
+                    .or_else(ConsensusGrpcError::into)
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply consensus/service/src/api/client_api_service.rs:329]: I think it's okay to just have two separate function calls, I think it is easier for the reader to follow "what happens when a user proposes a Tx".

We are also going to delete the v1 of this in the next release cycle, so this is not like long term technical debt.

File: fog/sample-paykit/src/client.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (garbageslam) [reply fog/sample-paykit/src/client.rs:182]: I was looking at how the test client actually gets this data, it looks that we already have a function for this that I missed. Going to refactor

File: consensus/enclave/impl/src/lib.rs [hunk 14]
```diff
@@ -1127,6 +1159,97 @@ mod tests {
...
+            .unwrap();
+
+        // First, create an encrypted connection from a client
+        let initiator = Start::new(responder_id.to_string());
+        let init_input = ClientInitiate::<X25519, Aes256Gcm, Sha512>::default();
+        let (initiator, auth_request_output) = initiator.try_next(&mut rng, init_input).unwrap();
+
+        let auth_request_msg = AuthMessage::from(auth_request_output);
+        let (auth_response_msg, _client_session) =
+            enclave.client_accept(auth_request_msg.into()).unwrap();
```
Comment (garbageslam) [inline consensus/enclave/impl/src/lib.rs:1187]: @jcape @nick-mobilecoin I tried to write a unit test per Eran's request that exercises the `client_tx_propose_v2` pathway. However, doing this requires creating an attested connection within the unit test, since that function takes an `attest::Message` argument.

I think I got that part right, but now I'm failing ike so:

```
running 1 test
test tests::test_client_tx_propose_works ... FAILED

failures:

---- tests::test_client_tx_propose_works stdout ----
thread 'tests::test_client_tx_propose_works' panicked at 'called `Result::unwrap()` on an `Err` value: Attest(NoReportAvailable)', consensus/enclave/impl/src/lib.rs:1193:103
```

I remember that we had some difficulties writing tests like this in the past, and may have decided not to write any of them because of this, since it will also all be influened by `SGX_MODE=HW` vs. `SGX_MODE=SW`, and I want to make sure I approach this in the best way.

* The only way to get the enclave to have a report is for it to accept a report via the reportable enclave API.
* I have to do the whole quoting enclave / RA client thing
* It seems like I can use the `mc-sgx-report-cache-untrusted` code in the unit test most likely, and then I think I only need to supply an RA client, and that might avoid some boilerplate.

The thing I'm worried about is that this unit test does not actually run in an enclave, so the `attest-ake` stuff may not work or may fail when it tries to create `Report::new()`, are those e-report instructions even going to work?

I remember that you said in the past that in simulation mode, the quoting enclave will accept anything. But I'm not sure if I can assume that this will run in simulation mode.

The other idea that I had was to go to `mc-crypto-attest-ake` and add a "test-only" feature gating a function that lets me inject a dummy `VerificationReport`. It seems like this is all going to get changed for DCAP anyways so maybe I shouldn't waste too much time on this.

LMK if you guys have thoughts.

I guess another approach would be to move the interesting bits of `client_tx_propose_v2` to a helper function that is called after the decryption happens, and then test that instead.

File: connection/src/traits.rs [hunk 15]
```diff
@@ -153,12 +154,27 @@ pub trait BlockchainConnection: Connection {
...
     /// Propose a transaction over the encrypted channel.
     /// Returns the number of blocks in the ledger at the time the call was
     /// received.
     fn propose_tx(&mut self, tx: &Tx) -> Result<u64>;
+
+    /// Propose a transaction over the encrypted channel (v2 API).
+    /// The user makes an assertion about what the current state of the minimum
+    /// fee map is, which the enclave checks and rejects the Tx if it is
+    /// different. This prevents an information leak identified in
+    /// TOB-MCCT-5
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply connection/src/traits.rs:176]: I'll find the link, gimme a bit

File: consensus/enclave/impl/src/lib.rs (hunk 14 above)
Comment (nick-mobilecoin) [reply consensus/enclave/impl/src/lib.rs:1187]: I doubt report will work outside of on enclave it will likely pull in SGX libs, which will then rely on enclave specific functions.

As long as there isn't a concern with info leakage I would vote for moving the encryption/decryption out to a higher level function that is excersized with an integration test, where the unit test focuses on the edge case and specifics of the behavior while the integration tests one happy path and one failure path (unless there is different handling for different failures).

Since we've been talking about testing levels, one of the criteria that is often used to define unit tests is "fast".  While there is no good criteria for _fast_, I usually think of single digit ms, or less, with a few outlier tests hitting 100ms or so.

This takes 4 seconds on my machine.  
```rust
        let mut ledger = create_ledger();
        let sender = AccountKey::random(&mut rng);
        let recipient = AccountKey::random(&mut rng);
        let n_blocks = 3;
        initialize_ledger(block_version, &mut ledger, n_blocks, &sender, &mut rng);
```

Comment (garbageslam) [reply consensus/enclave/impl/src/lib.rs:1187]: :+1:

the stuff where we have to create an LMDB database to get test data is some junk left behind by engineers who no longer work here.

Comment (garbageslam) [reply consensus/enclave/impl/src/lib.rs:1187]: I'm going to try to unit test this in `mc-consensus-enclave` instead. I don't feel super good about splitting up `client_propose_tx` into two functions just so that there is a private function that we can test. I think testing private functions has little value, I'd rather just do whatever I can to test the public API.

File: consensus/enclave/tests/enclave_api_tests.rs [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+use std::str::FromStr;
+
+lazy_static::lazy_static! {
+    pub static ref DUMMY_INT_GAUGE: IntGauge = IntGauge::new("foo".to_string(), "bar".to_string()).unwrap();
+}
+
+/// Test that we can exercise client_tx_propose_v2 and that it passes and fails
+/// as expected.
+#[test_with_logger]
+fn consensus_enclave_client_tx_propose_v2(logger: Logger) {
```
Comment (garbageslam) [inline consensus/enclave/tests/enclave_api_tests.rs:34]: I was able to create a proper test of the consensus enclave client_propose_tx_v2 this way

File: consensus/enclave/impl/src/lib.rs (hunk 14 above)
Comment (garbageslam) [reply consensus/enclave/impl/src/lib.rs]: gonna mark this resolved in favor of this approach: https://github.com/mobilecoinfoundation/mobilecoin/pull/2670#discussion_r997458095

**Comment by eranrund** [comment]
I'm still reviewing this, but I think adding the fee map digest into the `Tx` struct (as an `Option<Vec<u8>>`) is a  simpler change. I actually think that if `TxPrefix` was named `Tx` and `Tx` was named `ProposeTxRequest` (and contained a Tx and a signature), we wouldn't be having a discussion around this, it would just sit inside `ProposeTxRequest` alongside the actual Tx and its signature.

Sticking it in `Tx` is not perfect, but here are the pros and cons as I see them:

Pros:
- Less changes to clients: clients can now start including the fee_map_digest when they are ready to do so, and if it is there consensus will validate it. That means we can go and immediately add it to all the clients we control, without worrying about what consensus version we are running (since this new field will get ignored by current consensus). With the v2 API, we can't update the clients until we have deployed a version of consensus that has the v2 API, or we need to add forks in the client code (if block version is X then use V1, otherwise use V2). This is the main reason to consider this approach imo.
- By adding it to the Tx, we are going to need to pass it to `TxBuilder`, which means that clients building against the newer code will be forced to learn about this change. Otherwise, we can only hope they start using the V2 API (or learn about it the hard way when V1 is deprecated)
- Less code changes overall (at least I think that would be the outcome): No new GRPC API call, no new enclave API
- Fee map hash validation can become part of the Tx validation, and doesn't need a separate path of "in addition to validating the Tx data, here is some other stuff to validate separately but only validate once when we receive the Tx". Where this plays into account is not relevant right now, but imagine if we ever made fees dynamic, or stored the pending transaction queue. If we don't validate the fee map at the time we are considering the transaction for a block, it means that we have possibly validated it against a stale fee map. I don't know how much that matters (but in general I think this whole attack is pretty insignificant)

Cons:
- Including the fee map digest in the `Tx` means the `Tx` gets bigger, and this structure gets propagated to all consensus nodes, stored in precious enclaves memory, etc.
- Putting stuff inside the `Tx` structure is currently considered controversial because it's nice that it only contains a signature and signed data (see initial argument as to why I think these names are just poorly chosen)
- Clients wont explicitly know that they are calling a version of ProposeTx that validates the fee map digest (unless they rely on figuring that out based on MRENCLAVE or a block version change that happened at the same time this got deployed), and are also not forced to include this, so if they are doing their own transaction building, they might never start including it. This can be mitigated by making it required at some point (which is equivalent to removing the deprecated V1 api)

I don't feel strongly about either path, but figured I will put my thoughts out there for others to consider in case someone thinks one path is significantly superior to the other.

**Review by eranrund — APPROVED** [review]
Putting aside the discussion on whether this is the right approach or whether the fee_map_digest should be added to `Tx`, I think the code here is fine if we choose to go this route and not the other.

File: Cargo.lock [hunk 17]
```diff
@@ -1980,9 +1980,9 @@ checksum = "b294d6fa9ee409a054354afc4352b0b9ef7ca222c69b8812cbea9e7d2bf3783f"
 
 [[package]]
 name = "libc"
-version = "0.2.135"
+version = "0.2.134"
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply Cargo.lock:1983]: thanks, will fix

**Review by remoun — APPROVED** [review]

**Comment by garbageslam** [comment]
This is now moot, Eran fixed this a different way in a series of PRs linked from: https://github.com/mobilecoinfoundation/mobilecoin/issues/2611
